### ISSUE TITLE
Many details are not understood

### ISSUE THREAD

**Issue by GuilinF** [pytorch/tutorials#514]
i haven't understanded some detail code  in DQN code
，so i wish to getting some introduction and interpretation. The total code is as follow；

#!/usr/bin/env python
# -*- coding: utf-8 -*-
# author: Grimm time:2019/5/26
import gym
import math
import random
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
# namedtuple是继承自tuple的子类。
# nametuple创建一个和tuple类似的对象
# 而且对应拥有可访问的属性
# 可利用collcetions.namedtuple构建一个简单的类
from collections import namedtuple
from itertools import count
# 知道这个是调用的图像，具体的情况还得再看
from PIL import Image

import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
# 对PIL.Image进行变换
import torchvision.transforms as T

env = gym.make('CartPole-v0').unwrapped
# set up matplotlib
is_ipython = 'inline' in matplotlib.get_backend()
if is_ipython:
    from IPython import display
# 将matplotlib设置为交互方式
# 需要在.show()之前设置为ioff形式
plt.ion()

# if gpu is to be used
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

Transition = namedtuple(
    'Transition',
    ('state',
     'action',
     'next_state',
     'reward'))

class ReplayMemory(object):
    def __init__(self, capacity):
        self.capacity = capacity
        self.memory = []
        self.position = 0

    def push(self, *args):
        """Save a transition."""
        if len(self.memory) < self.capacity:
            # 如果自身的memory小于容量的话，给列表中添加空元素
            self.momery.append(None)
        self.memory[self.position] = Transition(*args)
        # 这部分代码没看懂，取余操作是得到的位置最大为capacity大小容量
        self.position = (self.position + 1) % self.capacity

    def sample(self, batch_size):
        # 相当于从self.memory中随机取出batch_size个数据
        return random.sample(self.memory, batch_size)

    def __len__(self):
        return len(self.memory)

class DQN(nn.Module):
    def __init__(self):
        super(DQN, self).__init__()
        # Conv2d二维卷积层(in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=True)
        # 输入信号通道、卷积产生通道、卷积核的尺寸，控制相关步骤的计算步长
        # 图像卷积函数，输入为3通道，输出为16张特征图，卷积核为5*5的正方形
        self.conv1 = nn.Conv2d(3, 16, kernel_size=5, stride=2)
        # 对小批量3d数据组成的4d输入进行批标准化操作
        # num_features, eps=1e-05, momentum=0.1, affine=True
        # 期望输入的特征数，给分母加上的值，动态均值和方差所使用的动量，true为给该层添加的可学习的仿射变换参数
        self.bn1 = nn.BatchNorm2d(16)
        self.conv2 = nn.Conv2d(16, 32, kernel_size=5, stride=2)
        self.bn2 = nn.BatchNorm2d(32)
        self.conv3 = nn.Conv2d(32, 32, kernel_size=5, stride=2)
        self.bn3 = nn.BatchNorm2d(32)
        # 对输入数据做线性变化,y=Ax+b
        # (in_features, out_features, bias=True)
        # 输入样本大小，输出样本大小，True学习偏置
        # 这个地方不懂，为什么要用448?？还是不懂
        self.head = nn.Linear(448, 2)

    def forward(self, x):
        # 一层卷积
        x = F.relu(self.bn1(self.conv1(x)))
        # 两层卷积
        x = F.relu(self.bn2(self.conv2(x)))
        # 三层卷积
        x = F.relu(self.bn3(self.conv3(x)))
        # 将第三次卷积的输出拉伸为一行,构成全连接层
        return self.head(x.view(x.size(0), -1))

# 这部分代码用于从环境中提取和处理图像
# 将多个transform组合起来使用
# ToPILImage()将tensor转化为Image()的操作
# torchvision.transforms.Resize(size, interpolation=2)
# Image.CUBIC为立方差值方式
resize = T.compose([T.ToPILImage(), T.Resize(
    40, interpolation=Image.CUBIC), T.ToTensor()])

# this is based on the code from gym
screen_width = 600

def get_cart_location():
    # x_threshold是指的什么阈值？
    world_width = env.x_threshold * 2
    scale = screen_width / world_width
    return int(env.state[0] * scale + screen_width / 2.0)

def get_screen():
    screen = env.render(mode='rgb_array').transpose(
        (2, 0, 1))  # transpose into torch order (CHW)
    # Strip off the top and bottom of the screen
    screen = screen[:, 160:320]
    view_width = 320
    cart_location = get_cart_location()
    if cart_location < view_width // 2:
        slice_range = slice(view_width)
    elif cart_location > (screen_width - view_width // 2):
        slice_range = slice(-view_width, None)
    else:
        slice_range = slice(cart_location - view_width // 2,
                            cart_location + view_width // 2)
    # Strip off the edges, so that we have a square image centered on a cart
    screen = screen[:, :, slice_range]
    # Convert to float, rescare, convert to torch tensor
    # (this doesn't require a copy)
    screen = np.ascontiguousarray(screen, dtype=np.float32) / 255
    screen = torch.from_numpy(screen)
    # Resize, and add a batch dimension (BCHW)
    return resize(screen).unsqueeze(0).to(device)

My questions are as follow:
1、self.head = nn.Linear(448, 2) in class DQN(nn.Module), Why the number 488 was chosen as the linear input? i need an  interpretation
2、screen = screen[:, 160:320] in def get_screen() ,i need an  interpretation
3、world_width = env.x_threshold * 2  in  def get_cart_location(), Is the x_threshold as  location threshold？i need an  interpretation
Thanks for everyone help

**Comment by jlin27** [comment]
Hi @GuilinF,

Thank you very much for reaching out.

- To help us answer question, could you link us to the specific tutorial where you're having these questions?
- In addition, I'd highly suggest posting your question in our PyTorch discussion forum where you'll get a quick response from our skilled community members. This github is more relevant for any issues with the tutorials repository. 

Hope to hear from you soon
